Improvements to rust target passing and rpm target architecture determination

anupdhml opened the pull request:
This PR first adds the ability to pass rust target from command-line, during rpm build.

Example:
```
cargo rpm build --target x86_64-unknown-linux-gnu
```

The target value gets used by cargo during project build. This is similar to the behavior of `target` as part of [package.metadata.rpm.cargo] from Cargo.toml (introduced as part of https://github.com/iqlusioninc/cargo-rpm/pull/24), but the commandline arg now allows rpm packaging for multiple rust targets without needing to edit Cargo.toml for every target.

When target is specified from both commandline and Cargo.toml, the value passed from commandline takes precedence -- the entry in Cargo.toml can be now taken a default target value for rpm packaging (and which can be overridden from commandline).

---

Secondly, this PR tries to automatically infer rpm target architecture from the rust target (either coming from Cargo.toml or the newly introduced command-line flag).

If rpm `target_architecture` is already specified from Cargo.toml, as part of [package.metadata.rpm] (ability introduced in https://github.com/iqlusioninc/cargo-rpm/pull/49), that value takes precedence.

Also added logging to make it clear how rpm target architecture gets set ultimately.

---

The changes here should resolve the issue https://github.com/iqlusioninc/cargo-rpm/issues/47 (if I'm understanding the need there correctly). 

Also at some point, `target_architecture` can probably be removed from [package.metadata.rpm] for configuration simplicity, if the strategy chosen here turns out to be sufficient for real-world needs.

Feel free to let me know anything else I should address here. Thanks!

anupdhml committed:
Add ability to pass rust target during rpm build

Example:
```
cargo rpm build --target x86_64-unknown-linux-gnu
```

The target value gets used by cargo during project build. This is similar
to the behavior of `target` as part of [package.metadata.rpm.cargo] from
Cargo.toml, but the commandline arg now allows rpm packaging for multiple
rust targets without needing to edit Cargo.toml for every target.

When target is specified from both commandline and Cargo.toml, the value
passed from commandline takes precedence -- the entry in Cargo.toml can
be now taken a default target value for rpm packaging (and which can be
overridden from commandline).

anupdhml committed:
Infer rpm target architecture from the rust target

If rpm target architecture is already specified from Cargo.toml (as part
of [package.metadata.rpm]), that value takes precedence.

Also added logging to make it clear how rpm target architecture gets set
ultimately.

rubdos commented:
> Secondly, this PR tries to automatically infer rpm target architecture from the rust target (either coming from Cargo.toml or the newly introduced command-line flag).
> 
> If rpm `target_architecture` is already specified from Cargo.toml, as part of [package.metadata.rpm] (ability introduced in #49), that value takes precedence.

FWIW, target arch can also be set through `.cargo/config`, but that's even more difficult I assume. I also believe that the `--target` flag should get precedence over the target in `Cargo.toml`; that way I can put the default target in Cargo.toml (which is in Git) and test-build for other archs without changing a committed file.

> The changes here should resolve the issue #47 (if I'm understanding the need there correctly).
> 
> Also at some point, `target_architecture` can probably be removed from [package.metadata.rpm] for configuration simplicity, if the strategy chosen here turns out to be sufficient for real-world needs.

Sounds like it does indeed! I'm testing out your change as I'm writing this in my CI for [Whisperfish](https://gitlab.com/rubdos/whisperfish/).

rubdos commented:
Update: seems to work for me. I'd still vote for `--target` overriding the default in `Cargo.toml` though, let me know what you think! :-)

Thanks for tackling this!

anupdhml commented:
> Update: seems to work for me. I'd still vote for `--target` overriding the default in `Cargo.toml` though, let me know what you think! :-)
> 
> Thanks for tackling this!

Not a problem -- we needed this feature too!

In this PR, `--target` does override the same field set as part of [package.metadata.rpm.cargo]. Or were you asking that rpm target architecture also be overridden by specified target (based on the automatic detection rules added here), even when `target_architecture` is present in [package.metadata.rpm]?

anupdhml commented:
@rubdos also thanks for testing this out from your CI!

rubdos commented:
> In this PR, `--target` does override the same field set as part of [package.metadata.rpm.cargo]. Or were you asking that rpm target architecture also be overridden by specified target (based on the automatic detection rules added here), even when `target_architecture` is present in [package.metadata.rpm]?

I misread your comment, sorry for that!

I just noticed that our build, that uses `i686-unknown-linux-gnu`, produces an RPM that claims to be i386. For our SailfishOS (basically armv7hl or i486 OpenSuse) application, we'd be using a i486 system, I'm unsure what would happen if we'd try to install `zypper in harbour-whisperfish` when only i386 packages are available.  Otoh, I was lying too, to zypper, because I provide RPM with an RPM target architecture i486 that *actually* contains a i686 executable *and* is linked with a i386 linker... Do you happen to know how to cleanly solve this, or is this solved on itself?

tony-iqlusion commented:
@anupdhml can you fix the conflicting files after merging #71? Thank you!

tony-iqlusion commented on src/config.rs:
```diff
@@ -209,3 +209,31 @@ pub fn append_rpm_metadata(
 }
+
+/// Get rpm target architecture based on the rust target
+pub fn get_target_architecture(target: &str) -> &str {
+    let mut parts = target.split('-');
+    let arch = parts.next().unwrap();
+    let abi = parts.last().unwrap_or("");
+
+    // based on the similar function from cargo-deb:
+    // https://github.com/mmstick/cargo-deb/blob/v1.23.1/src/manifest.rs#L909-L937
+    // with adjustments for valid values that `rpmbuild --target` takes
+    match (arch, abi) {
+        // https://doc.rust-lang.org/std/env/consts/constant.ARCH.html
+        // rustc --print target-list
+        // https://fedoraproject.org/wiki/Architectures
+        // https://github.com/rpm-software-management/rpm/blob/rpm-4.14.3-release/rpmrc.in#L156
+        ("mipsisa32r6", _) => "mipsr6",
+        ("mipsisa32r6el", _) => "mipsr6el",
+        ("mipsisa64r6", _) => "mips64r6",
+        ("mipsisa64r6el", _) => "mips64r6el",
+        ("powerpc", _) => "ppc",
+        ("powerpc64", _) => "ppc64",
+        ("powerpc64le", _) => "ppc64le",
+        ("x86", _) => "i386",
+        (arm, gnueabi) if arm.starts_with("arm") && gnueabi.ends_with("hf") => "armv7hl",
+        (arm, _) if arm.starts_with("arm") => "armv7l",
+        (other_arch, _) => other_arch,
+    }
+}
```
Perhaps it'd be worth having a type that handles this mapping?

anupdhml replied:
were you thinking of a new `target` struct that would be initialized based on the rust target string, with a method to get the rpm target architecture?

I wasn't sure where to place this function actually. Maybe it would be better to make it part of the `Builder` struct implementation in builder.rs, reusing the `target` already stored there. If you prefer that too, I can go ahead and make that change.

tony-iqlusion replied:
I was thinking of something more like `pub enum Target`, left open-ended with an `Other(String)` variant, with methods for e.g. parsing from a Rust target triple and outputting an `rpmbuild` target.

The `License` type is an example of the sort of idea in this crate:

https://github.com/iqlusioninc/cargo-rpm/blob/develop/src/license.rs

anupdhml replied:
ok got it. Will implement something similar when I get some time and post an update.

thanks for the example!

tony-iqlusion replied:
Awesome, thank you!

anupdhml replied:
Added a commit to address this. Let me know if it looks good to you!

anupdhml committed:
Make getting rpm dir and filename more robust

The function `get_rpm_dir_and_filename` handles path strings with no directory
elements (i.e. when it's just a filename). The set rpm output path is
always absolute currently so we don't hit this edge case, but extended the
function implementation for completeness.

Also documented relative/absolute path handling a bit better.

anupdhml commented:
> @anupdhml can you fix the conflicting files now that #71 has been merged? Thank you!

Done! also added a small follow-up to #71 as part of this mr -- let me know if you prefer that as a separate mr and I can break it out.

anupdhml commented:
> I just noticed that our build, that uses `i686-unknown-linux-gnu`, produces an RPM that claims to be i386. For our SailfishOS (basically armv7hl or i486 OpenSuse) application, we'd be using a i486 system, I'm unsure what would happen if we'd try to install `zypper in harbour-whisperfish` when only i386 packages are available. Otoh, I was lying too, to zypper, because I provide RPM with an RPM target architecture i486 that _actually_ contains a i686 executable _and_ is linked with a i386 linker... Do you happen to know how to cleanly solve this, or is this solved on itself?

For the `i686-unknown-linux-gnu` builds, are you getting rpms with `i386` in the name? Based on the changes on this branch, you should have received `i686` rpm.

In general, i686/i486 hosts will be able to run binaries built for i386 or i486 but the other way round won't work. For the rpm packaging (if you are concerned about other dependencies that your package needs). I am actually not sure if i386/i486 rpms can satisfy dependencies for an i686 package -- maybe it depends on how the deps are specified when building the rpm?

anupdhml committed:
Add dedicated type to handle rpm target architecture conversion

tony-iqlusion reviewed (approved):
Looks good now, thanks!

anupdhml commented:
awesome, thanks for the quick reviews!

rubdos commented:
> In general, i686/i486 hosts will be able to run binaries built for i386 or i486 but the other way round won't work. For the rpm packaging (if you are concerned about other dependencies that your package needs). I am actually not sure if i386/i486 rpms can satisfy dependencies for an i686 package -- maybe it depends on how the deps are specified when building the rpm?

Based on
```rust
             TargetArch::x86 => "i386",
```
I expected to get `i386` in the rpm's name, yes. Seems like I got `harbour-whisperfish-0.6.0-0.alpha.1.i686.rpm` instead, I was confused by this line in my build log:
```
+ rm -rf /builds/rubdos/whisperfish/target/i686-unknown-linux-gnu/release/rpmbuild/BUILDROOT/harbour-whisperfish-0.6.0-0.alpha.1.i386
```

Not sure how or why it decides to call the buildroot i386, but that might be the match arm I cited earlier.

Either way, your code does "the right thing" and it's just me that potentially has to lie to the OS. Sorry for the noise, thank you for the patch!
